File: python/substr.py
import sys

def longestSubstring(qstring: str) -> int:
    """Find the longest substring without repeating character,
    from the given string.
    This method uses the sliding window method.
    """
    parsed_char_set = set()
    leftIndex = 0
    longestLengthMatched = 0

    for rightIndex in range(len(qstring)):
        while qstring[rightIndex] in parsed_char_set:
            parsed_char_set.remove(qstring[leftIndex])
            leftIndex += 1
        parsed_char_set.add(qstring[rightIndex])
        longestLengthMatched = max(longestLengthMatched, rightIndex - leftIndex + 1)

    return longestLengthMatched

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: %s <string>" %(sys.argv[0]))
        sys.exit(1)

    print(longestSubstring(sys.argv[1]))

File: cpp/substr.cpp
#include<iostream>
#include<string>
#include<list>
#include<algorithm>

int longestSubstring(const std::string& str) {
    int maxLengthMatched = 0;
    int rightIndex = 0;
    std::list<char> char_set;
    for (char c : str) {
        auto it = std::find(char_set.begin(), char_set.end(), c);
        if (it != char_set.end()) {
            char_set.erase(char_set.begin(), std::next(it));
        }
        char_set.push_back(c);
        rightIndex++;
        int currMatch = char_set.size();
        if (currMatch > maxLengthMatched)
            maxLengthMatched = currMatch;
    }
    return maxLengthMatched;
}

#ifndef __UNITTEST__
int main(int argc, char *argv[]) {
    if (argc != 2) {
        std::cout << "Usage: " << argv[0] << " <string>" << std::endl;
        exit(1);
    }
    int len = longestSubstring(std::string(argv[1]));
    std::cout << len << std::endl;
    return 0;
}
#endif  // __UNITTEST__

